[QA-1571] retry iam update when group "does not exist"

andy7i opened the pull request:
fixes Issue # 2 from https://broadworkbench.atlassian.net/browse/QA-1571?src=confmacro



Example of the issue we're trying to retry on:
During workspace creation, updating the google iam on the project sometimes fails, which seems to be a race condition.
```
[INFO] [11:08:40.133] [scala-execution-context-global-574] o.b.d.w.google.HttpGoogleIamDAO - retry-able operation failed: retries remain but predicate failed, not retrying
com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
POST https://cloudresourcemanager.googleapis.com/v1/projects/terra-quality-9bc8e9d5:setIamPolicy
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "message" : "Group policy-c9a8c0f1-bb5a-40a7-b685-c7746aa10254@quality.firecloud.org does not exist.",
    "reason" : "badRequest"
  } ],
  "message" : "Group policy-c9a8c0f1-bb5a-40a7-b685-c7746aa10254@quality.firecloud.org does not exist.",
  "status" : "INVALID_ARGUMENT"
}

```

This PR adds a retry predicate for this scenario. 

We can't depend on google to propagate changes everywhere. Even if we try waiting until we can _get_ the group, it doesn't guarantee that we will be able to update that group in the very next call. So this was changed to take an optional parameter to retry. Instead of retrying from Rawls, we're adding the retry predicate here so that others can also use this. 



**PR checklist**
- [ ] For each library you've modified here, decide whether it requires a major, minor, or no version bump. (Click [here](/broadinstitute/workbench-libs/blob/develop/CONTRIBUTING.md) for guidance)

If you're doing a **major** or **minor** version bump to any libraries:
- [ ] Bump the version in `project/Settings.scala` `createVersion()`
- [ ] Update `CHANGELOG.md` for those libraries
- [ ] I promise I used `@deprecated` instead of deleting code where possible

In all cases:
- [ ] Replace the appropriate version hashes in `README.md` and the `CHANGELOG.md` for any libs you changed with `TRAVIS-REPLACE-ME` to auto-update the version string
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green (CI _and_ coverage tests)
- [ ] Squash commits and **merge to develop**
- [ ] Delete branch after merge

andy7i committed:
retry on google group "does not exist" messages.

Example:

```
[INFO] [11:08:40.133] [scala-execution-context-global-574] o.b.d.w.google.HttpGoogleIamDAO - retry-able operation failed: retries remain but predicate failed, not retrying
com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
POST https://cloudresourcemanager.googleapis.com/v1/projects/terra-quality-9bc8e9d5:setIamPolicy
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "message" : "Group policy-c9a8c0f1-bb5a-40a7-b685-c7746aa10254@quality.firecloud.org does not exist.",
    "reason" : "badRequest"
  } ],
  "message" : "Group policy-c9a8c0f1-bb5a-40a7-b685-c7746aa10254@quality.firecloud.org does not exist.",
  "status" : "INVALID_ARGUMENT"
}

```

andy7i committed:
tests fail due to a scalafmtCheckAll error

```

[error] 1 files must be formatted
[error] (workbenchGoogle / Compile / scalafmtCheck) 1 files must be formatted
```

andy7i commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleUtilities.scala:
```diff
@@ -110,6 +120,16 @@ object GoogleUtilities {
       res.getOrElse(false)
     }
+
+    private def compareErrorMessage(ex: GoogleJsonResponseException)(fn: String => Boolean): Boolean = {
```
follows the pattern of the above private functions

marctalbott reviewed (approved):
Is there a corresponding Rawls PR?

gpolumbo-broad commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleIamDAO.scala:
```diff
@@ -211,13 +211,15 @@ class HttpGoogleIamDAO(appName: String, googleCredentialMode: GoogleCredentialMo
+      whenInvalidValueOnBucketCreation,
+      whenNonHttpIOException,
+      when409,
+      when400WithDoesNotExistMessage
```
I know we had talked about this before, but I'm not sure how I feel about this approach now.  On one hand, making a request to `modifyIamRoles` or `getProjectPolicy` should legitimately return a `400` if the group/policy does not exist.  

However, as written, if the group/policy legitimately does not exist, this won't just fail and return, and instead it will always retry until it times out.  That's what we want for this `createWorkspace` bug we're trying to address, but that might not be what we want globally.  The reality is, this method looks to actually only be used during `createWorkspace`, at least in Rawls, but this might get used elsewhere in the future (or it might be already, I did not do a comprehensive search through everything that uses `workbench-libs`).

In other words this, "If Google thinks the group doesn't exist, I don't believe them!" logic should be isolated to place where we feel that way.  I think 99% of the time, if Google says the group doesn't exist, we should believe them.  

So maybe the better answer here is to put some retry logic in `createWorkspace`, after we call `syncPolicies`, we actually query Google and say, "I just told you to make this group, does it exist yet?" and we retry that until it succeeds (or times out) and then proceed with the `createWorkspace` flow after Google finally notices that the group exists.

andy7i replied:
great points. I think that does seem like a better approach instead of making changes in a shared library like this.

andy7i replied:
We can't depend on google to propagate changes everywhere. Even if we try waiting until we can _get_ the group, it doesn't guarantee that we will be able to update that group in the very next call. So this was changed to take an optional parameter to retry. Instead of retrying from Rawls, we're adding the retry predicate here so that others can also use this.

gpolumbo-broad reviewed (changes requested):
Sorry, my comment is a major change to overall solution.  If you feel strongly that your implementation is the right way to go or my concerns are actually not so bad, that's fine, just let me know.

andy7i committed:
We can't depend on google to propagate changes so this will make the update more resilient, as an optional parameter.

andy7i committed:
changelog

andy7i committed:
hopefully fix compilation error. (it compiles on my machine but not in github)

```
[error] /home/runner/work/workbench-libs/workbench-libs/google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleIamDAO.scala:227:28: missing argument list for method when400WithDoesNotExistMessage in object RetryPredicates
[error] Unapplied methods are only converted to functions when a function type is expected.
[error] You can make this conversion explicit by writing `when400WithDoesNotExistMessage _` or `when400WithDoesNotExistMessage(_)` instead of `when400WithDoesNotExistMessage`.
[error]       basePredicateList :+ when400WithDoesNotExistMessage
[error]                            ^
```

andy7i committed:
wat

marctalbott commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleIamDAO.scala:
```diff
@@ -192,48 +192,72 @@ class HttpGoogleIamDAO(appName: String, googleCredentialMode: GoogleCredentialMo
-          whenNonHttpIOException,
-          when409
+
+    val when400WithDoesNotExistMessagePredicateList: Seq[(Throwable => Boolean)] = Seq(when400WithDoesNotExistMessage)
```
nit: I'd prefer to skip this seq and append `when400WithDoesNotExistMessage` directly to the `basePredicateList` down in that if block.

andy7i replied:
that's what I had originally, but the PR tests do not compile..

andy7i committed:
Revert "wat"

This reverts commit 37c68b5c

andy7i committed:
Revert "hopefully fix compilation error. (it compiles on my machine but not in github)"

This reverts commit 1c0c241a

andy7i committed:
needed parentheses around whenGroupDoesNotExist and type declaration..

andy7i committed:
add test to make codecov happy
